From 9d7f7477e01443272af9f636cdb3c7542362e9bc Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Sat, 29 Oct 2022 10:02:33 +0800
Subject: [PATCH 071/595] [Huawei] cgroup: Destroy the link between memcg and
 blkcg on css_create() error path

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187773

-------------------------------

In css_create(), ss->css_alloc() has attached memcg to blkcg->memcg_list
if css is memcg_css. Therefore.@err_free_css tag, we should destroy this
link before calling rcu to free css, otherwise a UAF issue will occur
when accessing blkcg->memcg_list.

==================================================================
BUG: KASAN: use-after-free in __list_add_valid+0x2c/0x134 lib/list_debug.c:27
Read of size 8 at addr ffff30f6860730e8 by task syz-executor.2/7075

CPU: 2 PID: 7075 Comm: syz-executor.2 Not tainted 5.10.0-01387-ge7e15c9be34a #1
Hardware name: linux,dummy-virt (DT)
Call trace:
 dump_backtrace+0x0/0x40c arch/arm64/kernel/stacktrace.c:132
 show_stack+0x30/0x40 arch/arm64/kernel/stacktrace.c:196
 __dump_stack lib/dump_stack.c:77 [inline]
 dump_stack+0x1b4/0x248 lib/dump_stack.c:118
 print_address_description.constprop.0+0x28/0x4bc mm/kasan/report.c:390
 __kasan_report mm/kasan/report.c:550 [inline]
 kasan_report+0x110/0x210 mm/kasan/report.c:567
 check_memory_region_inline mm/kasan/generic.c:187 [inline]
 __asan_load8+0x98/0xc0 mm/kasan/generic.c:253
 __list_add_valid+0x2c/0x134 lib/list_debug.c:27
 __list_add include/linux/list.h:67 [inline]
 list_add include/linux/list.h:86 [inline]
 wb_attach_memcg_to_blkcg+0x340/0xa40 mm/backing-dev.c:1107
 mem_cgroup_css_alloc+0x3a4/0x6f0 mm/memcontrol.c:5735
 css_create+0x7c/0x4a0 kernel/cgroup/cgroup.c:5333
 cgroup_apply_control_enable+0x288/0x340 kernel/cgroup/cgroup.c:3140
 cgroup_mkdir+0x228/0x540 kernel/cgroup/cgroup.c:5560
 kernfs_iop_mkdir+0xbc/0xf4 fs/kernfs/dir.c:1134
 vfs_mkdir+0x2b8/0x3f0 fs/namei.c:3651
 do_mkdirat+0x210/0x23c fs/namei.c:3674
 __do_sys_mkdirat fs/namei.c:3685 [inline]
 __se_sys_mkdirat fs/namei.c:3683 [inline]
 __arm64_sys_mkdirat+0x50/0x60 fs/namei.c:3683
 __invoke_syscall arch/arm64/kernel/syscall.c:36 [inline]
 invoke_syscall arch/arm64/kernel/syscall.c:48 [inline]
 el0_svc_common.constprop.0+0xf4/0x414 arch/arm64/kernel/syscall.c:155
 do_el0_svc+0x50/0x11c arch/arm64/kernel/syscall.c:217
 el0_svc+0x20/0x30 arch/arm64/kernel/entry-common.c:353
 el0_sync_handler+0xe4/0x1e0 arch/arm64/kernel/entry-common.c:369
 el0_sync+0x148/0x180 arch/arm64/kernel/entry.S:683

Allocated by task 7059:
 kasan_save_stack+0x24/0x50 mm/kasan/common.c:48
 kasan_set_track mm/kasan/common.c:56 [inline]
 set_alloc_info mm/kasan/common.c:486 [inline]
 __kasan_kmalloc mm/kasan/common.c:518 [inline]
 __kasan_kmalloc.constprop.0+0xd4/0x130 mm/kasan/common.c:489
 kasan_kmalloc+0xc/0x1c mm/kasan/common.c:532
 __kmalloc+0x244/0x980 mm/slub.c:4024
 kmalloc include/linux/slab.h:568 [inline]
 kzalloc include/linux/slab.h:675 [inline]
 mem_cgroup_alloc+0x34/0x494 mm/memcontrol.c:5616
 mem_cgroup_css_alloc+0xf8/0x6f0 mm/memcontrol.c:5682
 css_create+0x7c/0x4a0 kernel/cgroup/cgroup.c:5333
 cgroup_apply_control_enable+0x288/0x340 kernel/cgroup/cgroup.c:3140
 cgroup_mkdir+0x228/0x540 kernel/cgroup/cgroup.c:5560
 kernfs_iop_mkdir+0xbc/0xf4 fs/kernfs/dir.c:1134
 vfs_mkdir+0x2b8/0x3f0 fs/namei.c:3651
 do_mkdirat+0x210/0x23c fs/namei.c:3674
 __do_sys_mkdirat fs/namei.c:3685 [inline]
 __se_sys_mkdirat fs/namei.c:3683 [inline]
 __arm64_sys_mkdirat+0x50/0x60 fs/namei.c:3683
 __invoke_syscall arch/arm64/kernel/syscall.c:36 [inline]
 invoke_syscall arch/arm64/kernel/syscall.c:48 [inline]
 el0_svc_common.constprop.0+0xf4/0x414 arch/arm64/kernel/syscall.c:155
 do_el0_svc+0x50/0x11c arch/arm64/kernel/syscall.c:217
 el0_svc+0x20/0x30 arch/arm64/kernel/entry-common.c:353
 el0_sync_handler+0xe4/0x1e0 arch/arm64/kernel/entry-common.c:369
 el0_sync+0x148/0x180 arch/arm64/kernel/entry.S:683

Freed by task 799:
 kasan_save_stack+0x24/0x50 mm/kasan/common.c:48
 kasan_set_track+0x24/0x34 mm/kasan/common.c:56
 kasan_set_free_info+0x24/0x44 mm/kasan/generic.c:361
 __kasan_slab_free.part.0+0x114/0x1a0 mm/kasan/common.c:470
 __kasan_slab_free mm/kasan/common.c:476 [inline]
 kasan_slab_free+0x2c/0x40 mm/kasan/common.c:477
 slab_free_hook mm/slub.c:1569 [inline]
 slab_free_freelist_hook mm/slub.c:1608 [inline]
 slab_free mm/slub.c:3179 [inline]
 kfree+0x12c/0xc10 mm/slub.c:4176
 __mem_cgroup_free+0xc0/0xd4 mm/memcontrol.c:5596
 mem_cgroup_free mm/memcontrol.c:5602 [inline]
 mem_cgroup_css_free+0x158/0x200 mm/memcontrol.c:5828
 css_free_rwork_fn+0x7c/0x630 kernel/cgroup/cgroup.c:5146
 process_one_work+0x474/0xb20 kernel/workqueue.c:2280
 worker_thread+0x120/0x92c kernel/workqueue.c:2426
 kthread+0x20c/0x25c kernel/kthread.c:313
 ret_from_fork+0x10/0x18 arch/arm64/kernel/entry.S:1034

Last call_rcu():
 kasan_save_stack+0x24/0x50 mm/kasan/common.c:48
 kasan_record_aux_stack+0xcc/0x110 mm/kasan/generic.c:349
 __call_rcu+0xc4/0x2a0 kernel/rcu/tree.c:2989
 call_rcu+0x14/0x20 kernel/rcu/tree.c:3063
 queue_rcu_work+0x88/0xe0 kernel/workqueue.c:1761
 css_create+0x1a8/0x4a0 kernel/cgroup/cgroup.c:5374
 cgroup_apply_control_enable+0x288/0x340 kernel/cgroup/cgroup.c:3140
 cgroup_mkdir+0x228/0x540 kernel/cgroup/cgroup.c:5560
 kernfs_iop_mkdir+0xbc/0xf4 fs/kernfs/dir.c:1134
 vfs_mkdir+0x2b8/0x3f0 fs/namei.c:3651
 do_mkdirat+0x210/0x23c fs/namei.c:3674
 __do_sys_mkdirat fs/namei.c:3685 [inline]
 __se_sys_mkdirat fs/namei.c:3683 [inline]
 __arm64_sys_mkdirat+0x50/0x60 fs/namei.c:3683
 __invoke_syscall arch/arm64/kernel/syscall.c:36 [inline]
 invoke_syscall arch/arm64/kernel/syscall.c:48 [inline]
 el0_svc_common.constprop.0+0xf4/0x414 arch/arm64/kernel/syscall.c:155
 do_el0_svc+0x50/0x11c arch/arm64/kernel/syscall.c:217
 el0_svc+0x20/0x30 arch/arm64/kernel/entry-common.c:353
 el0_sync_handler+0xe4/0x1e0 arch/arm64/kernel/entry-common.c:369
 el0_sync+0x148/0x180 arch/arm64/kernel/entry.S:683

Second to last call_rcu():
 kasan_save_stack+0x24/0x50 mm/kasan/common.c:48
 kasan_record_aux_stack+0xcc/0x110 mm/kasan/generic.c:349
 __call_rcu+0xc4/0x2a0 kernel/rcu/tree.c:2989
 call_rcu+0x14/0x20 kernel/rcu/tree.c:3063
 queue_rcu_work+0x88/0xe0 kernel/workqueue.c:1761
 css_release_work_fn+0x268/0x7fc kernel/cgroup/cgroup.c:5234
 process_one_work+0x474/0xb20 kernel/workqueue.c:2280
 worker_thread+0x120/0x92c kernel/workqueue.c:2426
 kthread+0x20c/0x25c kernel/kthread.c:313
 ret_from_fork+0x10/0x18 arch/arm64/kernel/entry.S:1034

The buggy address belongs to the object at ffff30f686072000
 which belongs to the cache kmalloc-8k of size 8192
The buggy address is located 4328 bytes inside of
 8192-byte region [ffff30f686072000, ffff30f686074000)
The buggy address belongs to the page:
page:fffffec3d9f81c00 refcount:1 mapcount:0 mapping:0000000000000000 index:0x0 pfn:0x146070
head:fffffec3d9f81c00 order:3 compound_mapcount:0 compound_pincount:0
flags: 0x17ffff800010200(slab|head|node=0|zone=2|lastcpupid=0xfffff)
raw: 017ffff800010200 fffffec3d95af408 fffffec3d95fe408 ffff30f64000e100
raw: 0000000000000000 0000000000010001 00000001ffffffff 0000000000000000
page dumped because: kasan: bad access detected

Memory state around the buggy address:
 ffff30f686072f80: fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
 ffff30f686073000: fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
>ffff30f686073080: fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
                                                          ^
 ffff30f686073100: fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
 ffff30f686073180: fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb fb
==================================================================

Memcg is attached to a blkcg->memcg_list in mem_cgroup_css_alloc();
when css_create() run failed after mem_cgroup_css_alloc(),
the memcg will be freed later without cgroup_mutex lock.
Therefore the memcg will be freed but memcg_node is not moved from
blkcg->memcg_list. In this case, create a new memcg, it will be uaf in
list_add() of wb_attach_memcg_blkcg() function.

To fix this problem, we add wb_kill_memcg_blkcg() when css_create() run
failed after mem_cgroup_css_alloc().

Fixes: 13b8199f0a47 ("cgroup: support cgroup writeback on cgroupv1")
Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 kernel/cgroup/cgroup.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/kernel/cgroup/cgroup.c b/kernel/cgroup/cgroup.c
index 25fe37e6d21f..055cf43458f5 100644
--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -5370,6 +5370,9 @@ static struct cgroup_subsys_state *css_create(struct cgroup *cgrp,
 	list_del_rcu(&css->sibling);
 err_free_css:
 	list_del_rcu(&css->rstat_css_node);
+#ifdef CONFIG_CGROUP_WRITEBACK
+	wb_kill_memcg_blkcg(css);
+#endif
 	INIT_RCU_WORK(&css->destroy_rwork, css_free_rwork_fn);
 	queue_rcu_work(cgroup_destroy_wq, &css->destroy_rwork);
 	return ERR_PTR(err);
-- 
2.17.1

